src/NHibernate/Type/CollectionType.cs

```diff
@@ -391,14 +392,36 @@ public virtual object InstantiateResult(object original)
 			// in which case we should never ever ever reset the dirty flag
 			// on the target because we simply do not know...
 			IPersistentCollection originalPc = original as IPersistentCollection;
-			IPersistentCollection resultPc = result as IPersistentCollection;
-			if(originalPc != null && resultPc!=null)
+			IPersistentCollection resultPc = target as IPersistentCollection;
+			if(resultPc!=null)
 			{
-				if (!originalPc.IsDirty)
+				if (elementsEqual || (originalPc != null && !originalPc.IsDirty))
```

Comment:
You are correct I am sure. To be honest, I don't know these parts of NH that well. I did a fix that solved my test cases and saw that all tests were green.

Can you "translate" your doubts into a failing test? Or can you point me in the right direction here? Seems there are some tests missing here...